File: server/src/main/scala/kpn/shared/data/Relation.scala
package kpn.shared.data

import kpn.shared.Timestamp
import kpn.shared.data.raw.RawRelation

case class Relation(raw: RawRelation, members: Seq[Member]) extends Element {

  def id: Long = raw.id

  def version: Int = raw.version

  def timestamp: Timestamp = raw.timestamp

  def changeSetId: Long = raw.changeSetId

  def tags: Tags = raw.tags

  override def isRelation: Boolean = true

  def nodeMembers: Seq[NodeMember] = {
    members.flatMap {
      case nodeMember: NodeMember => Some(nodeMember)
      case _ => None
    }
  }

  def wayMembers: Seq[WayMember] = {
    members.flatMap {
      case wayMember: WayMember => Some(wayMember)
      case _ => None
    }
  }

  def relationMembers: Seq[RelationMember] = {
    members.flatMap {
      case relationMember: RelationMember => Some(relationMember)
      case _ => None
    }
  }

  def relationMember(id: Long): RelationMember = {
    relationMembers.find(m => m.relation.id == id).get
  }
}

File: backend/api/src/main/java/kpn/api/common/Relation.java
package kpn.api.common;

import kpn.api.common.data.Element;
import kpn.api.common.data.Member;
import kpn.api.custom.Tag;
import kpn.api.custom.Timestamp;

import com.google.common.collect.ImmutableList;

public record Relation(
  Long id,
  Long version,
  Timestamp timestamp,
  Long changeSetId,
  ImmutableList<Tag> tags,
  ImmutableList<Member> members
) implements Element {
}

/* TODO migrate
package kpn.api.common

import kpn.api.common.data.Element
import kpn.api.common.data.Member
import kpn.api.common.data.Way
import kpn.api.common.data.raw.RawRelation
import kpn.api.custom.Tag
import kpn.api.custom.Timestamp

case class Relation(
  id: Long,
  version: Long,
  timestamp: Timestamp,
  changeSetId: Long,
  tags: Seq[Tag],
  members: Seq[Member]
) extends Element {

  override def isRelation: Boolean = true

  def nodeMembers: Seq[Member] = {
    members.filter(_.isNode)
  }

  def wayMembers: Seq[Member] = {
    members.filter(_.isWay)
  }

  def relationMembers: Seq[Member] = {
    members.filter(_.isRelation)
  }

  def relationIdMembers: Seq[Member] = {
    members.filter(_.isRelationId)
  }

  def relationMember(id: Long): Member = {
    relationMembers.find(m => m.relation.map(_.id).contains(id)).get
  }

  def ways: Seq[Way] = {
    wayMembers.flatMap(_.way)
  }

  def toRaw: RawRelation = {
    RawRelation(
      id,
      version,
      timestamp,
      changeSetId,
      members.map(_.toRaw),
      tags
    )
  }
}

*/
